Lib/ntpath.py

```diff
@@ -640,23 +638,68 @@ def commonpath(paths):
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        is_str = isinstance(filename, str)
+        if not is_str:
+            filename = os.fsdecode(filename)
+        extended_path_prefix = '\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_path_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved if is_str else os.fsencode(resolved)
+
+    def _extended_to_normal(path):
+        letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        if (len(path) > 6 and
+            path[4].upper() in letters and
+            path[5] == ':' and
+            path[6] == '\\'):
+            # extended path with \\?\ prefix
+            # 4 is len('\\?\')
+            normal_path = normpath(path[4:])
+        elif len(path) > 7 and path[:8].upper() == '\\\\?\\UNC\\':
+            # UNC path with \\?\ prefix - drop prefix
+            # 7 is len('\\?\UNC')
+            normal_path = normpath('\\' + path[7:])
+        else:
+            # not a UNC or drive-letter path
+            # return path as-is
+            return path
 
+        if len(normal_path) < MAX_PATH and normal_path == _getfullpathname(normal_path):
+            return normpath(normal_path)
+        return path
+except ImportError:
+    def realpath(filename):
+        filename = os.fspath(filename)
+        extended_path_prefix = '\\\\?\\' if isinstance(filename, str) else b'\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        return filename if is_extended_path else abspath(filename)
+
+    def _getfinalpathname(path):
```

Comment:
I wonder why this fallback was ever defined for non-Windows platforms. Maybe it was a mistake or left over from code that's long gone. I think we don't need this.

Reply:
Dropping it is easy however at glance `ntpath` module is intended to be working even on non-Windows platform - there are a few places that try to gracefully handle missing native implementations: 
- [missing _getvolumepathname](https://github.com/python/cpython/blob/master/Lib/ntpath.py#L259)
- [missing _getfullpathname](https://github.com/python/cpython/blob/master/Lib/ntpath.py#L520)

Reply:
Ah, I've misunderstood you - assumed you are talking about both fallbacks defined in case if importing `_getfinalpathname ` fails.